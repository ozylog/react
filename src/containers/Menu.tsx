import React, { useState } from 'react';
import { useContextSelector } from 'use-context-selector';
import styled from 'styled-components';

import Layout from './Layout';

import LayoutContext from './../contexts/LayoutContext';

const MenuContainer = styled.div`
  width: 400px;
  position: fixed;
  top: 0;
  left: 0;
  height: 100%;
  border-right: 1px solid #999;
  background: #fff;
`;

function Menu(): React.ReactElement | null {
  const [isMobile, setIsMobile] = useState(false);
  const isMenuExpanded = useContextSelector(LayoutContext, (state) => state.isMenuExpanded);

  const handleResize = () => {
    if (window.innerWidth < 800 ) {
      if (!isMobile) setIsMobile(!isMobile);
    } else {
      if (isMobile) setIsMobile(!isMobile);
    }
  };


  React.useEffect(() => {
    window.addEventListener('resize', handleResize);
    handleResize();

    return () => {
      window.removeEventListener('resize', handleResize);
    };
  }, []);

  const style = isMobile ? { width: '100%', borderRight: 0 } : undefined;

  if (isMenuExpanded) {
    return (
      <MenuContainer style={style}>
        <Layout title='Menu'>
          Menu
        </Layout>
      </MenuContainer>

    );
  }

  return null;

}

export default React.memo(Menu);
