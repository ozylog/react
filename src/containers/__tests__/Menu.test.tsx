import renderer from 'react-test-renderer';
import Menu from '../Menu';
import 'jest-styled-components';
import * as ContextSelector from 'use-context-selector';

describe('<Menu />', () => {
  it('should render not Menu component when isMenuExpanded = false', () => {
    const component = renderer.create(
      <Menu />
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  })

  it('should render Menu component when isMenuExpanded = true', () => {
    const useContextSelector = jest.spyOn(ContextSelector, 'useContextSelector');
    useContextSelector.mockReturnValueOnce(true);

    const component = renderer.create(
      <Menu />
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    useContextSelector.mockClear();
  })
});
