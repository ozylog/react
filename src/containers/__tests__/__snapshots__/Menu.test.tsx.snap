// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Menu /> should render Menu component when isMenuExpanded = true 1`] = `null`;

exports[`<Menu /> should render not Menu component when isMenuExpanded = false 1`] = `null`;
